tools/lambda-promtail/lambda-promtail/promtail.go

```diff
@@ -0,0 +1,166 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/prometheus/common/model"
+)
+
+const (
+	timeout    = 5 * time.Second
+	minBackoff = 100 * time.Millisecond
+	maxBackoff = 30 * time.Second
+	maxRetries = 10
+
+	reservedLabelTenantID = "__tenant_id__"
+
+	userAgent = "lambda-promtail"
+)
+
+type entry struct {
+	labels model.LabelSet
+	entry  logproto.Entry
+}
+
+type batch struct {
+	streams map[string]*logproto.Stream
+}
+
+func newBatch(entries ...entry) *batch {
+	b := &batch{
+		streams: map[string]*logproto.Stream{},
+	}
+
+	for _, entry := range entries {
+		b.add(entry)
+	}
+
+	return b
+}
+
+func (b *batch) add(e entry) {
+	labels := labelsMapToString(e.labels, reservedLabelTenantID)
+	if stream, ok := b.streams[labels]; ok {
+		stream.Entries = append(stream.Entries, e.entry)
+		return
+	}
+
+	b.streams[labels] = &logproto.Stream{
+		Labels:  labels,
+		Entries: []logproto.Entry{e.entry},
+	}
+}
+
+func labelsMapToString(ls model.LabelSet, without ...model.LabelName) string {
+	lstrs := make([]string, 0, len(ls))
+Outer:
+	for l, v := range ls {
+		for _, w := range without {
+			if l == w {
+				continue Outer
+			}
+		}
+		lstrs = append(lstrs, fmt.Sprintf("%s=%q", l, v))
+	}
+
+	sort.Strings(lstrs)
+	return fmt.Sprintf("{%s}", strings.Join(lstrs, ", "))
+}
+
+func (b *batch) encode() ([]byte, int, error) {
+	req, entriesCount := b.createPushRequest()
+	buf, err := proto.Marshal(req)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	buf = snappy.Encode(nil, buf)
+	return buf, entriesCount, nil
+}
+
+func (b *batch) createPushRequest() (*logproto.PushRequest, int) {
+	req := logproto.PushRequest{
+		Streams: make([]logproto.Stream, 0, len(b.streams)),
+	}
+
+	entriesCount := 0
+	for _, stream := range b.streams {
+		req.Streams = append(req.Streams, *stream)
+		entriesCount += len(stream.Entries)
+	}
+	return &req, entriesCount
+}
+
+func sendToPromtail(ctx context.Context, b *batch) error {
+	buf, _, err := b.encode()
```

Comment:
nit: personally I'd just inline the encode function

Reply:
I preferred to keep as close as what promtail do, but since we only use this once maybe it make sense
https://github.com/grafana/loki/blob/67a5c19b1717ba367e989d74778eedc375e60a26/clients/pkg/promtail/client/client.go#L294
https://github.com/grafana/loki/blob/67a5c19b1717ba367e989d74778eedc375e60a26/clients/pkg/promtail/client/batch.go#L95